### PR TITLE
Least Frequently Used cache eviction algorithm with complexity O(1)

### PR DESCRIPTION
**Pull request by mehmetalisavas**

### CODE REVIEW

File: lfu_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	if err == nil {
+		t.Fatal("test_key should not be in the cache")
+	}
+}
+
+//
+// BENCHMARK OPERATIONS
+//
+
+func BenchmarkLFUSet1000(b *testing.B) {
```
Comment (cihangir) [inline lfu_test.go]: you dont need "testXxxxNTimes" functions.

File: lfu_nots.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+	// size holds the limit of the LFU cache
+	size int
+
+	currentSize int
+}
+
+type cacheItem struct {
+	// key of cache value
+	k string
```
Comment (cihangir) [inline lfu_nots.go]: you dont need key here.

File: lfu_nots.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	return nil
+}
+
+// entry holds the frequency node informations
+type entry struct {
+	// freqCount holds the frequency number
+	freqCount int
+
+	// itemCount holds the items how many exist in list
+	listEntry map[*cacheItem]byte
```
Comment (cihangir) [inline lfu_nots.go]: instead of byte use struct{} it wont do memory allocation

File: lfu_nots.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+type cacheItem struct {
+	// key of cache value
+	k string
+
+	// value of cache value
+	v interface{}
+
+	// holds the frequency elements
+	freqElement *list.Element
```
Comment (cihangir) [inline lfu_nots.go]: could you explain why do you need to store "freqElement" here?

File: lfu_nots.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		if ci.freqElement == nil {
+			nextPosition = l.frequencyList.PushFront(entry)
+		} else {
+			nextPosition = l.frequencyList.InsertAfter(entry, ci.freqElement)
+		}
+	}
+	nextPosition.Value.(*entry).listEntry[ci] = 1
+	ci.freqElement = nextPosition
+
+	if ci.freqElement.Prev() != nil {
```
Comment (cihangir) [inline lfu_nots.go]: document why do you need to remove item from the list regardless of the list count.

File: lfu_nots.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+func (l *LFUNoTS) remove(ci *cacheItem, position *list.Element) {
+	entry := position.Value.(*entry).listEntry
+	delete(entry, ci)
+	if len(entry) == 0 {
+		l.frequencyList.Remove(position)
+	}
+}
+
+// newEntry creates a new entry with frequency count
+func newEntry(freqCount int) *entry {
```
Comment (cihangir) [inline lfu_nots.go]: this is not used.

**Comment by cihangir** [comment]
assign me back when you are done

File: lfu_nots.go (hunk 6 above)
Comment (mehmetalisavas) [inline lfu_nots.go]: https://github.com/koding/cache/pull/12/files#diff-1f45c95b8ef207e5ac27ea0f3d4cc76aR155

File: lfu_nots.go (hunk 2 above)
Comment (mehmetalisavas) [inline lfu_nots.go]: we hold the key-value pair together with its usage count inside freqElement in this struct. 
It looks like we need. Dont we ?

File: lfu_nots.go (hunk 4 above)
Comment (mehmetalisavas) [inline lfu_nots.go]: Documented

File: lfu_nots.go (hunk 5 above)
Comment (mehmetalisavas) [inline lfu_nots.go]: Documented

File: lfu_test.go (hunk 1 above)
Comment (mehmetalisavas) [inline lfu_test.go]: Done

File: lfu_nots.go (hunk 3 above)
Comment (mehmetalisavas) [inline lfu_nots.go]: changed as struct{}

**Comment by mehmetalisavas** [comment]
@cihangir last review ?

File: lfu_nots.go (hunk 2 above)
Comment (cihangir) [inline lfu_nots.go]: ok i didn't realize "evict" part, thx for the heads up

**Comment by cihangir** [comment]
Thanks!

**Pull request merged**